core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java

```diff
@@ -0,0 +1,411 @@
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * ForeignKey represents the foreign and unique key constraint relationship
+ * on the current {@link org.apache.calcite.rel.RelNode}.
+ *
+ * <p><b>constraints</b> field {@link #constraints} are
+ * constraints that foreign key and unique key relationships in bottom-up derivation.
+ *
+ * <p><b>foreignColumns</b> field {@link #foreignColumns} indicates the
+ * position of the foreign key on the current {@link org.apache.calcite.rel.RelNode}
+ * if not or not be confirmed, it is an empty set.
+ *
+ * <p><b>uniqueColumns</b> field {@link #uniqueColumns} indicates the position of
+ * the unique key on the current {@link org.apache.calcite.rel.RelNode},
+ * if not or not be confirmed, it is an empty set, the position in uniqueColumns
+ * corresponds to foreignColumns.
+ *
+ * <p>The element positions in {@code uniqueColumns} and {@code foreignColumns}
+ * correspond to each other. The order of elements in {@code uniqueColumns}
+ * and {@code foreignColumns} is consistent with the order of constraints in
+ * the constraints list.
+ *
+ * <p>For instance,
+ * <blockquote>
+ * <pre>select e.deptno, e.ename, d.deptno
+ * from emp as e
+ * inner join dept as d
+ * on e.deptno = d.deptno</pre></blockquote>
+ *
+ * <p>the foreign key is the DEPTNO column of CATALOG.SALES.EMP table,
+ * reference the DEPTNO unique column of CATALOG.SALES.DEPT table.
+ *
+ * <p>Invoke the {@link RelMetadataQuery#getConfirmedForeignKeys} method which
+ * input param is the top {@link org.apache.calcite.rel.core.Project},
+ * the following results can be obtained.
+ *
+ * <p>{@code constraints} is
+ * [{left: [CATALOG, SALES, EMP].$7.#true, right: [CATALOG, SALES, DEPT].$0.#true}]
+ * {@code foreignColumns} is {0}
+ * {@code uniqueColumns} is {2}
+ *
+ * <p>For instance,
+ * <blockquote>
+ * <pre>select name, deptno
+ * from dept</pre></blockquote>
+ *
+ * <p>{@code constraints} is
+ * [{left: null.null.#false, right: [CATALOG, SALES, DEPT].$0.#false}]
+ * {@code foreignColumns} is {}
+ * {@code uniqueColumns} is {1}
+ *
+ * <p>For instance,
+ * <blockquote>
+ * <pre>select ename, sal, deptno
+ * from emp</pre></blockquote>
+ *
+ * <p>{@code constraints} is
+ * [{left: [CATALOG, SALES, EMP].$7.#false, right: [CATALOG, SALES, DEPT].$0.#false}]
+ * {@code foreignColumns} is {2}
+ * {@code uniqueColumns} is {}
+ *
+ * @see InferredConstraintKey
+ * @see org.apache.calcite.plan.RelOptForeignKey
+ */
+public class RelOptForeignKey {
+
+  /** Foreign key and unique key relationships in bottom-up derivation. */
+  private final List<Pair<InferredConstraintKey, InferredConstraintKey>> constraints;
+  /** Position of the foreign key on the current {@link org.apache.calcite.rel.RelNode}. */
+  private final ImmutableBitSet foreignColumns;
+  /** Position of the unique key on the current {@link org.apache.calcite.rel.RelNode}. */
+  private final ImmutableBitSet uniqueColumns;
```

Comment:
Why is there information about a unique key in the RelOptForeignKey class?

Reply:
I think a foreign key is a reference to the Primary key of another table,make uniqueColumns to referredPrimaryKeys.

Make the name reflect its true meaning

WDYT?

Reply:
Add the referred prefix is more accessible, I will fix it.